src/resources.rs

```diff
@@ -2,6 +2,106 @@
 // SPDX-License-Identifier: Apache-2.0
 
 //! Structs to manage device resources.
+//!
+//! The high level flow of resource management among the VMM, the device manager, and the device
+//! is as below:
+//! 1) the VMM creates a new device object.
+//! 2) the VMM asks the new device object for its resource requirements.
+//! 3) the VMM allocates resources for the device object according to resource requirements.
+//! 4) the VMM passes the allocated resources to the device object.
+//! 5) the VMM registers the new device onto corresponding device managers according the allocated
+//!    resources.
+
+use std::{u16, u32, u64};
+
+/// Enumeration describing a device's resource requirements.
+pub enum DeviceResourceRequest {
+    /// Request for an IO Port address range.
+    PioAddress {
+        /// Reserve the preallocated IO ports [`fixed`, `fixed` + `size`] if it's not u16::MAX.
+        fixed: u16,
+        /// Allocate resource within the range [`min`, `max`] if `min` is not u16::MAX.
+        min: u16,
+        /// Allocate resource within the range [`min`, `max`] if `max` is not zero.
+        max: u16,
+        /// Alignment for the allocated address.
+        align: u16,
+        /// Size for the allocated address range.
+        size: u16,
+    },
+    /// Request for a Memory Mapped IO address range.
+    MmioAddress {
+        /// Reserve the preallocated range [`fixed`, `fixed` + `size`] if it's not u64::MAX.
+        fixed: u64,
+        /// Allocate resource within the range [`min`, `max`] if 'min' is not u64::MAX.
+        min: u64,
+        /// Allocate resource within the range [`min`, `max`] if 'max' is not zero.
+        max: u64,
+        /// Alignment for the allocated address.
+        align: u64,
+        /// Size for the allocated address range.
+        size: u64,
```

Comment:
It depends on the semantics the vm-allocator plans to support. Current definition follows the linux kernel  behavior. If we decide to simplify the vm-allocator semantics, the definition may be simplified too.

Reply:
@jiangliu I think @chao-p proposal would allow for representing all cases that we want to support, regardless of the vm-allocator semantics, unless I'm missing something. Also, this could be used for both the device request for resources and the allocated device resources as well. It simplifies a device resource description down to one enum.

Reply:
Seems it can't allocate address range constraints for 32-bit PCI BAR. Or am I missing some points?
A tuple (min, max, size, align) should satisfy our requirements, but not sure whether it could be reduced as three-ele tuple (base, size, align).

Reply:
> @jiangliu I think @chao-p proposal would allow for representing all cases that we want to support, regardless of the vm-allocator semantics, unless I'm missing something. 

FWIW, I think having a 32-bit `base` field implicitly means we want a 32-bit region. Same as the `max` field being set to `u32:MAX` in your current proposal.

Reply:
> > @jiangliu I think @chao-p proposal would allow for representing all cases that we want to support, regardless of the vm-allocator semantics, unless I'm missing something.
> 
> FWIW, I think having a 32-bit `base` field implicitly means we want a 32-bit region. Same as the `max` field being set to `u32:MAX` in your current proposal.

With the above proposal we would not be able to express a request for a region that would cross the 32-bit limit, but I don't think this is something that exists today.

Reply:
> > @jiangliu I think @chao-p proposal would allow for representing all cases that we want to support, regardless of the vm-allocator semantics, unless I'm missing something.
> 
> FWIW, I think having a 32-bit `base` field implicitly means we want a 32-bit region. Same as the `max` field being set to `u32:MAX` in your current proposal.

I got the point, you suggest to allocate MMIO address below 4G for 32-bit BAR and MMIO address  above 4G for 64-bit BARs. That means we need to maintain at least two MMIO address pools.
Currently firecracker maintains only one MMIO address pools below 4G, which serves both 32-bit and 64-bit BARs.

So I'm proposing a policy as:
1) allocate MMIO address below 4G for 32-bit BARs,
2) allocate MMIO address above 4G for 64-bit BARs, and fallback to MMIO address below 4G.

The difference is whether to support `fallback` case, which is the current firecracker design.

Reply:
> > > @jiangliu I think @chao-p proposal would allow for representing all cases that we want to support, regardless of the vm-allocator semantics, unless I'm missing something.
> > 
> > 
> > FWIW, I think having a 32-bit `base` field implicitly means we want a 32-bit region. Same as the `max` field being set to `u32:MAX` in your current proposal.
> 
> I got the point, you suggest to allocate MMIO address below 4G for 32-bit BAR and MMIO address above 4G for 64-bit BARs. That means we need to maintain at least two MMIO address pools.
>

With `cloud-hypervisor` we had to support directly assigned devices that had both 32-bit and 64-bit BARs. Initially we were planning to actually only have 64-bit BARs but that does not work when you have more than 3 BARs (Due to the PCI config space limits). The lessons we learned from that is that:

1. We must put 32-bit BARs below 4G, on the 32-bit hole.
2. Assuming the 32-bit hole will be spacious enough for dynamic environments (with hotplugged and VFIO devices) and can fit all devices allocations is risky.

So we do maintain 2 separate MMIO pools, yes. While I'm writing this, I realize the following proposed definition by @chao-p:

```Rust
MmioAddress {
    base: Option<u64>,
    size: u64,
    align: u64
}
```

Would not be able to express the "Give me *any* 32-bit MMIO region" but only "Give me a 32-bit MMIO region starting at 0xf00ba4". So I think it should expanded to include e.g. a bitflags to describe further requirements (e.g. the 32-bit one).

> Currently firecracker maintains only one MMIO address pools below 4G, which serves both 32-bit and 64-bit BARs.
> 
> So I'm proposing a policy as:
> 
> 1. allocate MMIO address below 4G for 32-bit BARs,
> 2. allocate MMIO address above 4G for 64-bit BARs, and fallback to MMIO address below 4G.
> 
> The difference is whether to support `fallback` case, which is the current firecracker design.

By fallback you mean using the 32-bit address space when there is no 64-bit one?

Reply:
> > > > @jiangliu I think @chao-p proposal would allow for representing all cases that we want to support, regardless of the vm-allocator semantics, unless I'm missing something.
> > > 
> > > 
> > > FWIW, I think having a 32-bit `base` field implicitly means we want a 32-bit region. Same as the `max` field being set to `u32:MAX` in your current proposal.
> > 
> > 
> > I got the point, you suggest to allocate MMIO address below 4G for 32-bit BAR and MMIO address above 4G for 64-bit BARs. That means we need to maintain at least two MMIO address pools.
> 
> With `cloud-hypervisor` we had to support directly assigned devices that had both 32-bit and 64-bit BARs. Initially we were planning to actually only have 64-bit BARs but that does not work when you have more than 3 BARs (Due to the PCI config space limits). The lessons we learned from that is that:
> 
> 1. We must put 32-bit BARs below 4G, on the 32-bit hole.
> 2. Assuming the 32-bit hole will be spacious enough for dynamic environments (with hotplugged and VFIO devices) and can fit all devices allocations is risky.
> 
> So we do maintain 2 separate MMIO pools, yes. While I'm writing this, I realize the following proposed definition by @chao-p:
> 
> ```rust
> MmioAddress {
>     base: Option<u64>,
>     size: u64,
>     align: u64
> }
> ```
> 
> Would not be able to express the "Give me _any_ 32-bit MMIO region" but only "Give me a 32-bit MMIO region starting at 0xf00ba4". So I think it should expanded to include e.g. a bitflags to describe further requirements (e.g. the 32-bit one).
> 
Adding flags essentially makes it a four-elements tuples, just different encoding method.
> > Currently firecracker maintains only one MMIO address pools below 4G, which serves both 32-bit and 64-bit BARs.
> > So I'm proposing a policy as:
> > 
> > 1. allocate MMIO address below 4G for 32-bit BARs,
> > 2. allocate MMIO address above 4G for 64-bit BARs, and fallback to MMIO address below 4G.
> > 
> > The difference is whether to support `fallback` case, which is the current firecracker design.
> 
> By fallback you mean using the 32-bit address space when there is no 64-bit one?
Yes, fallback means using 32-bit addresses for 64-bit BARs.

Reply:
> > > > > @jiangliu I think @chao-p proposal would allow for representing all cases that we want to support, regardless of the vm-allocator semantics, unless I'm missing something.
> > > > 
> > > > 
> > > > FWIW, I think having a 32-bit `base` field implicitly means we want a 32-bit region. Same as the `max` field being set to `u32:MAX` in your current proposal.
> > > 
> > > 
> > > I got the point, you suggest to allocate MMIO address below 4G for 32-bit BAR and MMIO address above 4G for 64-bit BARs. That means we need to maintain at least two MMIO address pools.
> > 
> > 
> > With `cloud-hypervisor` we had to support directly assigned devices that had both 32-bit and 64-bit BARs. Initially we were planning to actually only have 64-bit BARs but that does not work when you have more than 3 BARs (Due to the PCI config space limits). The lessons we learned from that is that:
> > 
> > 1. We must put 32-bit BARs below 4G, on the 32-bit hole.
> > 2. Assuming the 32-bit hole will be spacious enough for dynamic environments (with hotplugged and VFIO devices) and can fit all devices allocations is risky.
> > 
> > So we do maintain 2 separate MMIO pools, yes. While I'm writing this, I realize the following proposed definition by @chao-p:
> > ```rust
> > MmioAddress {
> >     base: Option<u64>,
> >     size: u64,
> >     align: u64
> > }
> > ```
> > 
> > 
> > Would not be able to express the "Give me _any_ 32-bit MMIO region" but only "Give me a 32-bit MMIO region starting at 0xf00ba4". So I think it should expanded to include e.g. a bitflags to describe further requirements (e.g. the 32-bit one).
> 
> Adding flags essentially makes it a four-elements tuples, just different encoding method.

That's correct. I think the flags option gives us more flexibility on how we could describe constraints. 
I'm trying to see what would be the more convenient way of using the same resource definition for both requested and reserved resources, for simplicity sake.
Using a definition like:
```Rust
MmioAddress {
    base: Option<u64>,
    flags: u64,
    size: u64,
    align: u64
}
```
seems, at least to me, a more natural way of expressing both a requested MMIO resource and a reserved one.

> > > Currently firecracker maintains only one MMIO address pools below 4G, which serves both 32-bit and 64-bit BARs.
> > > So I'm proposing a policy as:
> > > 
> > > 1. allocate MMIO address below 4G for 32-bit BARs,
> > > 2. allocate MMIO address above 4G for 64-bit BARs, and fallback to MMIO address below 4G.
> > > 
> > > The difference is whether to support `fallback` case, which is the current firecracker design.
> > 
> > 
> > By fallback you mean using the 32-bit address space when there is no 64-bit one?
> Yes, fallback means using 32-bit addresses for 64-bit BARs.

I believe we should keep that as an allocator strategy. What the selected VMM allocator does with the following MMIO request:

```Rust
let mut flags = 1u64 << MMIO_64;
let requested_region = MMIOAddress {
         base: None,
         flags,
         size: 0x10000,
         align: 0x1000,
};

is up to the allocator itself. It could very well decide to return a 32-bit address from the 32-bit hole.

```